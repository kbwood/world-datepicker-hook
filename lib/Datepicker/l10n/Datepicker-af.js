"use strict";

var _Datepicker = require("../Datepicker");

/* http://keith-wood.name/world-datepicker.html
   Afrikaans localisation for world datepicker.
   Written by Renier Pretorius and Ruediger Thiede. */
_Datepicker.localisations.af = {
  clear: 'Vee uit',
  clearLabel: 'Verwyder die huidige datum',
  close: 'Klaar',
  closeLabel: 'Sluit sonder verandering',
  // current: 'Vandag',
  // currentLabel: 'Vertoon huidige maand',
  dayLabel: 'Kies DD, M d',
  defaultLabel: 'Kies \'n datum',
  isRTL: false,
  monthLabel: 'Vertoon \'n ander maand',
  nextMonth: 'Volgende',
  nextMonthLabel: 'Vertoon volgende maand',
  nextYear: '>>',
  nextYearLabel: 'Vertoon volgende jaar',
  prevMonth: 'Vorige',
  prevMonthLabel: 'Vertoon vorige maand',
  prevYear: '<<',
  prevYearLabel: 'Vertoon vorige jaar',
  today: 'Vandag',
  todayLabel: 'Vertoon huidige maand',
  week: 'Wk',
  yearLabel: 'Vertoon \'n ander jaar',
  weekLabel: 'Week van die jaar'
};
//# sourceMappingURL=Datepicker-af.js.map