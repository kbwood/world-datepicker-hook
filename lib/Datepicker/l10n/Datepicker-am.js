"use strict";

var _Datepicker = require("../Datepicker");

/* http://keith-wood.name/world-datepicker.html
   Amharic (አማርኛ) localisation for world datepicker.
   Leyu Sisay. */
_Datepicker.localisations.am = {
  clear: 'አጥፋ',
  clearLabel: 'የተመረጠውን ቀን አጥፋ',
  close: 'ዝጋ',
  closeLabel: 'የቀን መምረጫውን ዝጋ',
  // current: 'አሁን',
  // currentLabel: 'የአሁኑን ወር አሳይ',
  dayLabel: 'DD, M d, yyyy ምረጥ',
  defaultLabel: 'ቀን ምረጥ',
  isRTL: false,
  monthLabel: 'ወሩን ቀይር',
  nextMonth: 'ቀጣይ',
  nextMonthLabel: 'ቀጣዩን ወር አሳይ',
  nextYear: '>>',
  nextYearLabel: 'ቀጣዩን ዓመት አሳይ',
  prevMonth: 'ያለፈ',
  prevMonthLabel: 'ያለፈውን ወር አሳይ',
  prevYear: '<<',
  prevYearLabel: 'ያለፈውን ዓመት አሳይ',
  today: 'ዛሬ',
  todayLabel: 'የዛሬን ወር አሳይ',
  week: 'ሳም',
  weekLabel: 'የዓመቱ ሳምንት ',
  yearLabel: 'ዓመቱን ቀይር'
};
//# sourceMappingURL=Datepicker-am.js.map