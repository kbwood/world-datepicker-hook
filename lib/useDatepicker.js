"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = require("react");

var _worldCalendars = _interopRequireDefault(require("@kbwood/world-calendars"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var generateCurrent = curDate => {
  var calendar = curDate.calendar();
  var yearString = "".concat(curDate.year());
  return {
    date: curDate,
    month: curDate.month(),
    monthName: calendar.local.monthNames[curDate.month() - calendar.firstMonth],
    monthNameShort: calendar.local.monthNamesShort[curDate.month() - calendar.firstMonth],
    year: curDate.year(),
    yearLocal: calendar.local.localiseDigits ? calendar.local.localiseDigits(yearString) : yearString
  };
};

var generateDays = (onSelect, curDate, setCurDate, minDate, maxDate, _ref) => {
  var {
    selectDaysInOtherMonths,
    showDaysInOtherMonths
  } = _ref;
  var calendar = curDate.calendar();

  var localiseDigits = calendar.local.localiseDigits || (value => value);

  var today = calendar.date();
  var daysInWeek = calendar.daysInWeek();
  var monthFirst = curDate.set(calendar.minDay, 'd');
  var monthCalc = monthFirst.sub(monthFirst.dayOfWeek(), 'd').add(calendar.local.firstDay, 'd');
  var monthStart = monthCalc.month() === curDate.month() && monthCalc.day() > calendar.minDay ? monthCalc.sub(1, 'w') : monthCalc;
  var monthLast = monthFirst.add(1, 'm').sub(1, 'd');
  var weekCount = Math.ceil((monthLast.toJD() - monthStart.toJD() + 1) / daysInWeek);
  var monthEnd = monthStart.add(weekCount, 'w').sub(1, 'd');
  var days = [];
  var forDay = monthStart;
  var week = 0;

  var _loop = function _loop() {
    var thisDay = forDay.date();
    var isInCurrentMonth = thisDay.month() === curDate.month();
    var isDisabled = !!minDate && thisDay.compareTo(minDate) === -1 || !!maxDate && thisDay.compareTo(maxDate) === 1 || !(selectDaysInOtherMonths || isInCurrentMonth);
    var day = {
      date: thisDay,
      isDisabled,
      isInCurrentMonth,
      isSelected: thisDay.compareTo(curDate) === 0,
      isToday: thisDay.compareTo(today) === 0,
      isWeekend: !thisDay.weekDay(),
      label: showDaysInOtherMonths || isInCurrentMonth ? "".concat(localiseDigits(String(thisDay.day()))) : '',
      selectDay: () => {
        setCurDate(thisDay);
        onSelect(thisDay);
      }
    };

    if (!days[week]) {
      days[week] = [];
    }

    days[week].push(day);
    forDay = forDay.add(1, 'd');

    if (forDay.dayOfWeek() === calendar.local.firstDay) {
      week += 1;
    }
  };

  while (forDay.compareTo(monthEnd) <= 0) {
    _loop();
  }

  return days;
};

var generateYears = (curDate, minDate, maxDate, options) => {
  var years = [];

  if (!options.selectYear) {
    return years;
  }

  var yearLimits = (options.yearRange || '').split(':').map(value => parseInt(value, 10));

  if (yearLimits.length !== 2 || yearLimits.some(value => isNaN(value))) {
    return years;
  }

  yearLimits[0] = minDate ? Math.max(minDate.year(), yearLimits[0]) : yearLimits[0];
  yearLimits[1] = maxDate ? Math.min(maxDate.year(), yearLimits[1]) : yearLimits[1];
  var calendar = curDate.calendar();

  var localiseDigits = calendar.local.localiseDigits || (value => value);

  for (var y = yearLimits[0]; y <= yearLimits[1]; y += 1) {
    years.push({
      name: localiseDigits("".concat(y)),
      value: y
    });
  }

  return years;
};

var generateLocal = (curDate, minDate, maxDate, options) => {
  var calendar = curDate.calendar();
  var daysInWeek = calendar.daysInWeek();
  var year = curDate.year();
  var {
    dayNames,
    dayNamesMin,
    dayNamesShort,
    firstDay,
    isRTL,
    monthNames,
    monthNamesShort
  } = calendar.local;

  var isDisabled = m => !!maxDate && calendar.date(year, m, calendar.minDay).compareTo(maxDate) > 0 || !!minDate && calendar.date(year, m, calendar.daysInMonth(year, m)).compareTo(minDate) < 0;

  return {
    days: dayNames.map((_, i) => {
      var index = (i + firstDay) % daysInWeek;
      return {
        name: dayNames[index],
        nameMin: dayNamesMin[index],
        nameShort: dayNamesShort[index]
      };
    }),
    isRTL,
    months: monthNames.map((name, i) => ({
      disabled: isDisabled(i + calendar.firstMonth),
      name,
      nameShort: monthNamesShort[i],
      value: i + calendar.firstMonth
    })),
    years: generateYears(curDate, minDate, maxDate, options)
  };
};

var getStartOfWeek = date => {
  var firstDay = date.calendar().local.firstDay;
  var dow = date.dayOfWeek();
  return date.sub(dow < firstDay ? date.calendar().daysInWeek() : dow, 'd').add(firstDay, 'd');
};

var getEndOfWeek = date => getStartOfWeek(date).add(date.calendar().daysInWeek() - 1, 'd');

var max = function max(d1) {
  var d2 = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : d1;
  return d1.compareTo(d2) === 1 ? d1 : d2;
};

var min = function min(d1) {
  var d2 = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : d1;
  return d1.compareTo(d2) === -1 ? d1 : d2;
};

var generateUpdates = (curDate, setCurDate, minDate, maxDate) => {
  var setTarget = target => () => {
    setCurDate(max(min(target, maxDate), minDate));
  };

  return {
    nextDay: setTarget(curDate.add(1, 'd')),
    nextMonth: setTarget(curDate.add(1, 'm')),
    nextWeek: setTarget(curDate.add(1, 'w')),
    nextYear: setTarget(curDate.add(1, 'y')),
    prevDay: setTarget(curDate.sub(1, 'd')),
    prevMonth: setTarget(curDate.sub(1, 'm')),
    prevWeek: setTarget(curDate.sub(1, 'w')),
    prevYear: setTarget(curDate.sub(1, 'y')),
    setMonth: month => {
      setTarget(curDate.set(parseInt("".concat(month), 10), 'm'))();
    },
    setYear: year => {
      setTarget(curDate.set(parseInt("".concat(year), 10), 'y'))();
    },
    today: setTarget(curDate.calendar().date()),
    weekEnd: setTarget(getEndOfWeek(curDate)),
    weekStart: setTarget(getStartOfWeek(curDate))
  };
};

var generateDatepicker = (onSelect, curDate, setCurDate, minDate, maxDate, options) => ({
  current: generateCurrent(curDate),
  days: generateDays(onSelect, curDate, setCurDate, minDate, maxDate, options),
  local: generateLocal(curDate, minDate, maxDate, options),
  updates: generateUpdates(curDate, setCurDate, minDate, maxDate)
});

var noOptions = {};

var useDatepicker = _ref2 => {
  var {
    calendarName,
    date,
    language = '',
    maxDate,
    minDate,
    onSelect,
    options = noOptions
  } = _ref2;

  var calendar = _worldCalendars.default.instance(calendarName, language);

  var [curDate, setCurDate] = (0, _react.useState)(date || calendar.date());
  var [datepicker, setDatepicker] = (0, _react.useState)(generateDatepicker(onSelect, curDate, setCurDate, minDate, maxDate, options));
  (0, _react.useEffect)(() => {
    var newDate = date || calendar.date();

    if (newDate.calendar() !== curDate.calendar() || newDate.compareTo(curDate) !== 0) {
      setCurDate(newDate);
    }
  }, [calendar, date]);
  (0, _react.useEffect)(() => {
    setDatepicker(generateDatepicker(onSelect, curDate, setCurDate, minDate, maxDate, options));
  }, [curDate, maxDate, minDate, onSelect, options, setCurDate, setDatepicker]);
  return datepicker;
};

var _default = useDatepicker;
exports.default = _default;
//# sourceMappingURL=useDatepicker.js.map