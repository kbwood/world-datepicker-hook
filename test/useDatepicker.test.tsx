import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Calendars, { CDate } from '@kbwood/world-calendars';
import '@kbwood/world-calendars/lib/Gregorian';
import '@kbwood/world-calendars/lib/l10n/Gregorian-zh-CN';
import useDatepicker, { Day, DisplayOptions, LocalMonth, LocalYear } from '../src/useDatepicker';

type TestProps = {
  date?: CDate,
  language?: string,
  maxDate?: CDate,
  minDate?: CDate,
  options?: DisplayOptions,
}

describe('(Hook) useDatepicker', () => {
  const defaultOptions = { selectDaysInOtherMonths: true, showDaysInOtherMonths: true };
  const gregorian = Calendars.instance('gregorian');
  const user = userEvent.setup();
  const onSelect = jest.fn();

  const generateWeeks = (days: Day[][]) => (
    <tbody>
      {days.map(week => (
        <tr key={`w${week[0].date.toString()}`}>
          {week.map(day => {
            const {
              date,
              isDisabled,
              isInCurrentMonth,
              isSelected,
              isToday,
              isWeekend,
              label,
              selectDay
            } = day;
            const className = `${isInCurrentMonth ? '' : 'other'} ${
              isSelected ? 'selected' : ''
            } ${isToday ? 'today' : ''} ${isWeekend ? 'weekend' : ''}`;
            return (
              <td key={date.toString()} className={className}>
                <button
                  disabled={isDisabled}
                  onClick={() => { selectDay(); }}
                  tabIndex={isSelected ? 0 : -1}
                  type="button"
                >
                  {label}
                </button>
              </td>
            );
          })}
        </tr>
      ))}
    </tbody>
  );

  const generateMonths = (months: LocalMonth[]) => (
    <select>
      {months.map(({ disabled, name }) => <option key={name} disabled={disabled}>{name}</option>)}
    </select>
  );

  const generateYears = (years: LocalYear[]) => (
    <select>
      {years.map(({ name }) => <option key={name}>{name}</option>)}
    </select>
  );

  const TestDatepicker = ({ date, language, maxDate, minDate, options = defaultOptions }: TestProps) => {
    const { current, days, local, updates } =
      useDatepicker({ calendarName: 'Gregorian', date, language, maxDate, minDate, onSelect, options });
    return (
      <>
        <div>
          <button onClick={() => { updates.prevYear(); }} type="button">
            Prev year
          </button>
          <button onClick={() => { updates.prevMonth(); }} type="button">
            Prev month
          </button>
          <button onClick={() => { updates.prevWeek(); }} type="button">
            Prev week
          </button>
          <button onClick={() => { updates.prevDay(); }} type="button">
            Prev day
          </button>
          <button onClick={() => { updates.weekStart(); }} type="button">
            Start week
          </button>
          <button onClick={() => { updates.today(); }} type="button">
            Today
          </button>
          <button onClick={() => { updates.weekEnd(); }} type="button">
            End week
          </button>
          <button onClick={() => { updates.nextDay(); }} type="button">
            Next day
          </button>
          <button onClick={() => { updates.nextWeek(); }} type="button">
            Next week
          </button>
          <button onClick={() => { updates.nextMonth(); }} type="button">
            Next month
          </button>
          <button onClick={() => { updates.nextYear(); }} type="button">
            Next year
          </button>
          <button onClick={() => { updates.setMonth('5'); }} type="button">
            Set month
          </button>
          <button onClick={() => { updates.setYear('2020'); }} type="button">
            Set year
          </button>
        </div>
        <table role="grid">
          <thead>
            <tr key="month">
              <th colSpan={local.days.length}>
                {options.selectMonth ? generateMonths(local.months) : current.monthName}{' '}
                {options.selectYear ? generateYears(local.years) : current.yearLocal}
              </th>
            </tr>
            <tr key="days">
              {local.days.map(({ name, nameMin }) => (
                <th key={name}>{nameMin}</th>
              ))}
            </tr>
          </thead>
          {generateWeeks(days)}
        </table>
      </>
    );
  };

  beforeAll(() => {
    const now = new Date(2022, 7 - 1, 13, 12, 0, 0, 0).getTime();
    jest.spyOn(Date, 'now').mockImplementation(() => now);
  });

  afterAll(() => {
    jest.restoreAllMocks();
  });

  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe('Default', () => {
    it('should show the current month', () => {
      render(<TestDatepicker />);

      expect(screen.getByRole('columnheader', { name: 'July 2022' })).toBeInTheDocument();
      expect(screen.getByRole('row', { name: 'Su Mo Tu We Th Fr Sa' })).toBeInTheDocument();
      expect(screen.getByRole('cell', { name: '13' })).toHaveClass('selected');
      expect(screen.getAllByRole('row')[2]).toMatchSnapshot();
    });

    it('should not show days in other months', () => {
      render(<TestDatepicker options={{ showDaysInOtherMonths: false }}/>);

      expect(screen.getAllByRole('row')[2]).toMatchSnapshot();
    });

    it('should allow month selection when requested', () => {
      render(<TestDatepicker options={{ selectMonth: true }}/>);
      const select = screen.getByRole('combobox') as HTMLSelectElement;

      expect(select).toBeInTheDocument();
      expect(select.options.length).toBe(12);
      for (let i = 0; i < select.options.length; i += 1) {
        const option = select.options.item(i);
        expect((option || {}).text).toBe(gregorian.local.monthNames[i]);
      }
    });

    it('should limit month selection when restricted by min/max date', () => {
      render(
        <TestDatepicker
          maxDate={gregorian.date(2022, 10, 13)}
          minDate={gregorian.date(2022, 3, 13)}
          options={{ selectMonth: true }}
        />
      );
      const select = screen.getByRole('combobox') as HTMLSelectElement;

      expect(select).toBeInTheDocument();
      expect(select.options.length).toBe(12);
      for (let i = 0; i < select.options.length; i += 1) {
        const option = select.options.item(i);
        expect(option).not.toBeNull();
        expect((option || {}).text).toBe(gregorian.local.monthNames[i]);
        expect((option || {}).disabled).toBe(i < 2 || i > 9);
      }
    });

    it('should allow year selection when requested', () => {
      render(<TestDatepicker options={{ selectYear: true, yearRange: '2020:2025' }}/>);
      const select = screen.getByRole('combobox') as HTMLSelectElement;

      expect(select).toBeInTheDocument();
      expect(select.options.length).toBe(6);
      for (let i = 0; i < select.options.length; i += 1) {
        const option = select.options.item(i);
        expect((option || {}).text).toBe(`${i + 2020}`);
      }
    });

    it('should limit year selection when restricted by min/max date', () => {
      render(
        <TestDatepicker
          maxDate={gregorian.date(2023, 10, 13)}
          minDate={gregorian.date(2021, 3, 13)}
          options={{ selectYear: true, yearRange: '2020:2025' }}
        />
      );
      const select = screen.getByRole('combobox') as HTMLSelectElement;

      expect(select).toBeInTheDocument();
      expect(select.options.length).toBe(3);
      for (let i = 0; i < select.options.length; i += 1) {
        const option = select.options.item(i);
        expect((option || {}).text).toBe(`${i + 2021}`);
      }
    });

    it('should not display years for selection when empty range', () => {
      render(<TestDatepicker options={{ selectYear: true }}/>);
      const select = screen.getByRole('combobox') as HTMLSelectElement;

      expect(select).toBeInTheDocument();
      expect(select.options.length).toBe(0);
    });

    it('should not display years for selection when only one range value', () => {
      render(<TestDatepicker options={{ selectYear: true, yearRange: '2020' }}/>);
      const select = screen.getByRole('combobox') as HTMLSelectElement;

      expect(select).toBeInTheDocument();
      expect(select.options.length).toBe(0);
    });

    it('should not display years for selection when invalid range', () => {
      render(<TestDatepicker options={{ selectYear: true, yearRange: '2020:sometime' }}/>);
      const select = screen.getByRole('combobox') as HTMLSelectElement;

      expect(select).toBeInTheDocument();
      expect(select.options.length).toBe(0);
    });

    it('should respond to moving to the previous year', async () => {
      render(<TestDatepicker />);
      await user.click(screen.getByRole('button', { name: 'Prev year' }));

      expect(screen.getByRole('columnheader', { name: 'July 2021' })).toBeInTheDocument();
      expect(screen.getByRole('cell', { name: '13' })).toHaveClass('selected');
    });

    it('should respond to moving to the previous month', async () => {
      render(<TestDatepicker />);
      await user.click(screen.getByRole('button', { name: 'Prev month' }));

      expect(screen.getByRole('columnheader', { name: 'June 2022' })).toBeInTheDocument();
      expect(screen.getByRole('cell', { name: '13' })).toHaveClass('selected');
    });

    it('should respond to moving to the previous week', async () => {
      render(<TestDatepicker />);
      await user.click(screen.getByRole('button', { name: 'Prev week' }));

      expect(screen.getByRole('cell', { name: '13' })).not.toHaveClass('selected');
      expect(screen.getAllByRole('cell', { name: '6' })[0]).toHaveClass('selected');
    });

    it('should respond to moving to the previous day', async () => {
      render(<TestDatepicker />);
      await user.click(screen.getByRole('button', { name: 'Prev day' }));

      expect(screen.getByRole('cell', { name: '13' })).not.toHaveClass('selected');
      expect(screen.getByRole('cell', { name: '12' })).toHaveClass('selected');
    });

    it('should respond to moving to the start of the week', async () => {
      render(<TestDatepicker />);
      await user.click(screen.getByRole('button', { name: 'Start week' }));

      expect(screen.getByRole('cell', { name: '13' })).not.toHaveClass('selected');
      expect(screen.getByRole('cell', { name: '10' })).toHaveClass('selected');
    });

    it('should respond to moving to today', async () => {
      render(<TestDatepicker date={gregorian.date(2022, 7, 19)} />);

      expect(screen.getByRole('cell', { name: '13' })).not.toHaveClass('selected');
      expect(screen.getByRole('cell', { name: '19' })).toHaveClass('selected');

      await user.click(screen.getByRole('button', { name: 'Today' }));

      expect(screen.getByRole('cell', { name: '13' })).toHaveClass('selected');
      expect(screen.getByRole('cell', { name: '19' })).not.toHaveClass('selected');
    });

    it('should respond to moving to the end of the week', async () => {
      render(<TestDatepicker />);
      await user.click(screen.getByRole('button', { name: 'End week' }));

      expect(screen.getByRole('cell', { name: '13' })).not.toHaveClass('selected');
      expect(screen.getByRole('cell', { name: '16' })).toHaveClass('selected');
    });

    it('should respond to moving to the next day', async () => {
      render(<TestDatepicker />);
      await user.click(screen.getByRole('button', { name: 'Next day' }));

      expect(screen.getByRole('cell', { name: '13' })).not.toHaveClass('selected');
      expect(screen.getByRole('cell', { name: '14' })).toHaveClass('selected');
    });

    it('should respond to moving to the next week', async () => {
      render(<TestDatepicker />);
      await user.click(screen.getByRole('button', { name: 'Next week' }));

      expect(screen.getByRole('cell', { name: '13' })).not.toHaveClass('selected');
      expect(screen.getByRole('cell', { name: '20' })).toHaveClass('selected');
    });

    it('should respond to moving to the next month', async () => {
      render(<TestDatepicker />);
      await user.click(screen.getByRole('button', { name: 'Next month' }));

      expect(screen.getByRole('columnheader', { name: 'August 2022' })).toBeInTheDocument();
      expect(screen.getByRole('cell', { name: '13' })).toHaveClass('selected');
    });

    it('should respond to moving to the next year', async () => {
      render(<TestDatepicker />);
      await user.click(screen.getByRole('button', { name: 'Next year' }));

      expect(screen.getByRole('columnheader', { name: 'July 2023' })).toBeInTheDocument();
      expect(screen.getByRole('cell', { name: '13' })).toHaveClass('selected');
    });

    it('should respond to setting the month', async () => {
      render(<TestDatepicker />);
      await user.click(screen.getByRole('button', { name: 'Set month' }));

      expect(screen.getByRole('columnheader', { name: 'May 2022' })).toBeInTheDocument();
      expect(screen.getByRole('cell', { name: '13' })).toHaveClass('selected');
    });

    it('should respond to setting the year', async () => {
      render(<TestDatepicker />);
      await user.click(screen.getByRole('button', { name: 'Set year' }));

      expect(screen.getByRole('columnheader', { name: 'July 2020' })).toBeInTheDocument();
      expect(screen.getByRole('cell', { name: '13' })).toHaveClass('selected');
    });

    it('should respond to selecting a date', async () => {
      render(<TestDatepicker />);
      await user.click(screen.getByRole('button', { name: '19' }));

      expect(screen.getByRole('columnheader', { name: 'July 2022' })).toBeInTheDocument();
      expect(screen.getAllByRole('cell', { name: '1' })[0]).not.toHaveClass('selected');
      expect(screen.getByRole('cell', { name: '19' })).toHaveClass('selected');
      expect(onSelect).toHaveBeenCalledTimes(1);
      expect(onSelect).toHaveBeenCalledWith(gregorian.date(2022, 7, 19));
    });

    it('should respond to selecting a date in another month', async () => {
      render(<TestDatepicker />);
      await user.click(screen.getAllByRole('button', { name: '29' })[0]);

      expect(screen.getByRole('columnheader', { name: 'June 2022' })).toBeInTheDocument();
      expect(screen.getAllByRole('cell', { name: '29' })[0]).not.toHaveClass('selected');
      expect(screen.getAllByRole('cell', { name: '1' })[0]).not.toHaveClass('selected');
      expect(screen.getAllByRole('cell', { name: '29' })[1]).toHaveClass('selected');
      expect(onSelect).toHaveBeenCalledTimes(1);
      expect(onSelect).toHaveBeenCalledWith(gregorian.date(2022, 6, 29));
    });
  });

  describe('With min/max date', () => {
    it('should disable dates before the minimum date', () => {
      render(<TestDatepicker minDate={gregorian.date(2022, 7, 12)} />);

      expect(screen.getAllByRole('button', { name: '1' })[0]).toBeDisabled();
      expect(screen.getByRole('button', { name: '11' })).toBeDisabled();
      expect(screen.getByRole('button', { name: '12' })).not.toBeDisabled();
      expect(screen.getAllByRole('button', { name: '30' })[1]).not.toBeDisabled();
    });

    it('should disable dates after the maximum date', () => {
      render(<TestDatepicker maxDate={gregorian.date(2022, 7, 15)} />);

      expect(screen.getAllByRole('button', { name: '1' })[0]).not.toBeDisabled();
      expect(screen.getByRole('button', { name: '15' })).not.toBeDisabled();
      expect(screen.getByRole('button', { name: '16' })).toBeDisabled();
      expect(screen.getAllByRole('button', { name: '30' })[1]).toBeDisabled();
    });

    it('should respond to moving to the previous year when restricted by a minimum date', async () => {
      render(<TestDatepicker minDate={gregorian.date(2022, 3, 17)} />);
      await user.click(screen.getByRole('button', { name: 'Prev year' }));

      expect(screen.getByRole('columnheader', { name: 'March 2022' })).toBeInTheDocument();
      expect(screen.getByRole('cell', { name: '17' })).toHaveClass('selected');
      expect(screen.getByRole('button', { name: '16' })).toBeDisabled();
      expect(screen.getByRole('button', { name: '17' })).not.toBeDisabled();
    });

    it('should respond to moving to the previous month  when restricted by a minimum date', async () => {
      render(<TestDatepicker minDate={gregorian.date(2022, 6, 17)} />);
      await user.click(screen.getByRole('button', { name: 'Prev month' }));

      expect(screen.getByRole('columnheader', { name: 'June 2022' })).toBeInTheDocument();
      expect(screen.getByRole('cell', { name: '17' })).toHaveClass('selected');
      expect(screen.getByRole('button', { name: '16' })).toBeDisabled();
      expect(screen.getByRole('button', { name: '17' })).not.toBeDisabled();
    });

    it('should respond to moving to the previous week when restricted by a minimum date', async () => {
      render(<TestDatepicker minDate={gregorian.date(2022, 7, 8)} />);
      await user.click(screen.getByRole('button', { name: 'Prev week' }));

      expect(screen.getByRole('cell', { name: '13' })).not.toHaveClass('selected');
      expect(screen.getByRole('cell', { name: '8' })).toHaveClass('selected');
    });

    it('should respond to moving to the previous day when restricted by a minimum date', async () => {
      render(<TestDatepicker minDate={gregorian.date(2022, 7, 13)} />);
      await user.click(screen.getByRole('button', { name: 'Prev day' }));

      expect(screen.getByRole('cell', { name: '13' })).toHaveClass('selected');
    });

    it('should respond to moving to the start of the week when restricted by a minimum date', async () => {
      render(<TestDatepicker minDate={gregorian.date(2022, 7, 12)} />);
      await user.click(screen.getByRole('button', { name: 'Start week' }));

      expect(screen.getByRole('cell', { name: '13' })).not.toHaveClass('selected');
      expect(screen.getByRole('cell', { name: '12' })).toHaveClass('selected');
    });

    it('should respond to moving to the end of the week when restricted by a maximum date', async () => {
      render(<TestDatepicker maxDate={gregorian.date(2022, 7, 14)} />);
      await user.click(screen.getByRole('button', { name: 'End week' }));

      expect(screen.getByRole('cell', { name: '13' })).not.toHaveClass('selected');
      expect(screen.getByRole('cell', { name: '14' })).toHaveClass('selected');
    });

    it('should respond to moving to the next day when restricted by a maximum date', async () => {
      render(<TestDatepicker maxDate={gregorian.date(2022, 7, 13)} />);
      await user.click(screen.getByRole('button', { name: 'Next day' }));

      expect(screen.getByRole('cell', { name: '13' })).toHaveClass('selected');
    });

    it('should respond to moving to the next week when restricted by a maximum date', async () => {
      render(<TestDatepicker maxDate={gregorian.date(2022, 7, 18)} />);
      await user.click(screen.getByRole('button', { name: 'Next week' }));

      expect(screen.getByRole('cell', { name: '13' })).not.toHaveClass('selected');
      expect(screen.getByRole('cell', { name: '18' })).toHaveClass('selected');
    });

    it('should respond to moving to the next month when restricted by a maximum date', async () => {
      render(<TestDatepicker maxDate={gregorian.date(2022, 8, 7)} />);
      await user.click(screen.getByRole('button', { name: 'Next month' }));

      expect(screen.getByRole('columnheader', { name: 'August 2022' })).toBeInTheDocument();
      expect(screen.getByRole('cell', { name: '7' })).toHaveClass('selected');
    });

    it('should respond to moving to the next year when restricted by a maximum date', async () => {
      render(<TestDatepicker maxDate={gregorian.date(2023, 3, 17)} />);
      await user.click(screen.getByRole('button', { name: 'Next year' }));

      expect(screen.getByRole('columnheader', { name: 'March 2023' })).toBeInTheDocument();
      expect(screen.getByRole('cell', { name: '17' })).toHaveClass('selected');
    });
  });

  describe('Localised', () => {
    it('should show the current month', () => {
      render(<TestDatepicker language="zh-CN" />);

      expect(screen.getByRole('columnheader', { name: '七月 二千二十二' })).toBeInTheDocument();
      expect(screen.getByRole('row', { name: '一 二 三 四 五 六 日' })).toBeInTheDocument();
      expect(screen.getAllByRole('row')[2]).toMatchSnapshot();
    });
  });
});
