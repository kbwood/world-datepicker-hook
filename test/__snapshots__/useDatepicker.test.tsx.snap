// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Hook) useDatepicker Default should not show days in other months 1`] = `
<tr>
  <td
    class="other   weekend"
  >
    <button
      disabled=""
      tabindex="-1"
      type="button"
    />
  </td>
  <td
    class="other   "
  >
    <button
      disabled=""
      tabindex="-1"
      type="button"
    />
  </td>
  <td
    class="other   "
  >
    <button
      disabled=""
      tabindex="-1"
      type="button"
    />
  </td>
  <td
    class="other   "
  >
    <button
      disabled=""
      tabindex="-1"
      type="button"
    />
  </td>
  <td
    class="other   "
  >
    <button
      disabled=""
      tabindex="-1"
      type="button"
    />
  </td>
  <td
    class="   "
  >
    <button
      tabindex="-1"
      type="button"
    >
      1
    </button>
  </td>
  <td
    class="   weekend"
  >
    <button
      tabindex="-1"
      type="button"
    >
      2
    </button>
  </td>
</tr>
`;

exports[`(Hook) useDatepicker Default should show the current month 1`] = `
<tr>
  <td
    class="other   weekend"
  >
    <button
      tabindex="-1"
      type="button"
    >
      26
    </button>
  </td>
  <td
    class="other   "
  >
    <button
      tabindex="-1"
      type="button"
    >
      27
    </button>
  </td>
  <td
    class="other   "
  >
    <button
      tabindex="-1"
      type="button"
    >
      28
    </button>
  </td>
  <td
    class="other   "
  >
    <button
      tabindex="-1"
      type="button"
    >
      29
    </button>
  </td>
  <td
    class="other   "
  >
    <button
      tabindex="-1"
      type="button"
    >
      30
    </button>
  </td>
  <td
    class="   "
  >
    <button
      tabindex="-1"
      type="button"
    >
      1
    </button>
  </td>
  <td
    class="   weekend"
  >
    <button
      tabindex="-1"
      type="button"
    >
      2
    </button>
  </td>
</tr>
`;

exports[`(Hook) useDatepicker Localised should show the current month 1`] = `
<tr>
  <td
    class="other   "
  >
    <button
      tabindex="-1"
      type="button"
    >
      二十七
    </button>
  </td>
  <td
    class="other   "
  >
    <button
      tabindex="-1"
      type="button"
    >
      二十八
    </button>
  </td>
  <td
    class="other   "
  >
    <button
      tabindex="-1"
      type="button"
    >
      二十九
    </button>
  </td>
  <td
    class="other   "
  >
    <button
      tabindex="-1"
      type="button"
    >
      三十
    </button>
  </td>
  <td
    class="   "
  >
    <button
      tabindex="-1"
      type="button"
    >
      一
    </button>
  </td>
  <td
    class="   weekend"
  >
    <button
      tabindex="-1"
      type="button"
    >
      二
    </button>
  </td>
  <td
    class="   weekend"
  >
    <button
      tabindex="-1"
      type="button"
    >
      三
    </button>
  </td>
</tr>
`;
